x-pack/plugins/lens/public/xy_visualization_plugin/xy_suggestions.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { partition } from 'lodash';
+import { Position } from '@elastic/charts';
+import { XYArgs } from './xy_expression';
+import { SuggestionRequest, VisualizationSuggestion, TableColumn } from '../types';
+
+/**
+ * Generate suggestions for the xy chart.
+ *
+ * @param opts
+ */
+export function getSuggestions(
+  opts: SuggestionRequest<XYArgs>
+): Array<VisualizationSuggestion<XYArgs>> {
+  return Object.entries(opts.tableColumns)
+    .filter(
+      ([_id, columns]) =>
+        // We require a column for x and one for y, and y needs numeric columns...
+        columns.length > 1 && columns.find(x => x.operation.dataType === 'number')
+    )
+    .map(([id, columns]) => getSuggestionForColumns(id, columns));
+}
+
+function getSuggestionForColumns(
+  id: string,
+  columns: TableColumn[]
+): VisualizationSuggestion<XYArgs> {
+  const [buckets, values] = partition(prioritizeColumns(columns), x => x.operation.isBucketed);
+
+  if (buckets.length >= 1) {
+    const [x, splitBy] = buckets;
+    return getSuggestion(id, x, values, splitBy);
+  } else {
+    const [x, ...yValues] = values;
+    return getSuggestion(id, x, yValues);
+  }
+}
+
+// This shuffles columns around so that the left-most column defualts to:
+// date, string, boolean, then number, in that priority. We then use this
+// order to pluck out the x column, and the split / stack column.
+function prioritizeColumns(columns: TableColumn[]) {
+  const columnSortOrder = {
+    date: 0,
+    string: 1,
+    boolean: 2,
+    number: 3,
+  };
+
+  return [...columns].sort(
+    (a, b) => columnSortOrder[a.operation.dataType] - columnSortOrder[b.operation.dataType]
```

Comment:
Looks like if we add a new data type, this would be NaN?

Reply:
Yeah. I was wondering what we should do in that case. At first, I thought we should throw an exception, but I thought better of it. We could just default any unknown types to be left of 'number' in the list. Thoughts?

Reply:
Another possibility would be to bail out and not suggest anything if there are unknown columns, because the result might probably not be what the user expects.

Reply:
Hm. It's tough. I lean towards suggesting something, since suggestions will probably be the most common way users get into a chart. Even if we suggest something slightly wonky, they can correct it, or we can file a bug and fix it explicitly... Not sure. @wylieconlon thoughts?

Reply:
Hm, I see what you mean. But in some cases the results could not make sense at all Example: the user pulls in a geo field and the datasource suggests geogrid on the geo field and count as a metric. The xy chart will happily try to display the geogrid value on the x axis which will probably render something like “[object Object]” as labels because there are geoobjects in the datatable. Maybe we are OK with that, not sure

Reply:
I'm also in favor of not making a suggestion unless the visualization understands all parts of it. We don't want to be like Clippy and make bad suggestions, so it's better to return only good suggestions.